### DISCUSSION TITLE
How do I preview documents with Tailwind?

### DISCUSSION THREAD

**Discussion by younes-io** [quarto-dev/quarto-cli#8259] [Q&A] [answered]
### Description

I want to use tailwind instead of Bootstrap in all the app, so I used `include-in-header`. I used the [example ](https://tailwindcss.com/docs/utility-first#why-not-just-use-inline-styles) from the Tailwind docs:

```
---
title: "Fake app"
pagetitle: "Fake page title"
subtitle: "Fake role"

format:
  html:
    include-in-header:
      - text: |
          <script src="https://cdn.jsdelivr.net/npm/tailwindcss@3.4.1/lib/index.min.js"></script>
          <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/tailwindcss@3.4.1/base.min.css">
---



::: {.py-8 .px-8 .max-w-sm .mx-auto .bg-white .rounded-xl .shadow-lg .space-y-2 .sm:py-4 .sm:flex .sm:items-center .sm:space-y-0 .sm:space-x-6}
<img class="block mx-auto h-24 rounded-full sm:mx-0 sm:shrink-0" src="/img/erin-lindford.jpg" alt="Woman's Face" />

::: {.text-center .space-y-2 .sm:text-left}
::: {.space-y-0.5}
Erin Lindford
::: {.text-lg .text-black .font-semibold}
Product Engineer
::: {.text-slate-500 .font-medium}
:::

<button class="px-4 py-1 text-sm text-purple-600 font-semibold rounded-full border border-purple-200 hover:text-white hover:bg-purple-600 hover:border-transparent focus:outline-none focus:ring-2 focus:ring-purple-600 focus:ring-offset-2">Message</button>
:::
:::
```

This is the output

![image](https://github.com/quarto-dev/quarto-cli/assets/3153107/953f5405-2e5f-4162-8e0c-b4fc84b921ca)

**Comment by cscheid** [comment]
```
: {.text-lg .text-black .font-semibold}
```

What do you expect this to do?

**Reply by younes-io** [reply]
Sorry it was a typo from another snippet.

**Reply by younes-io** [reply]
I updated the post.. sorry for the typo..

**Comment by mcanouil** [comment]
@younes-io could you edit the title of your post to actually include your question? Than you.

**Reply by younes-io** [reply]
Done, thank you

**Comment by mcanouil** [comment]
The code shared is incorrect, you did not even close all the divs (5 opened, 3 closed) and you should not use raw HTML outside of raw inline/block elements (` ```{=html} ` or  `` `TEXT`{=html} ``).
Also divs should be surrounded by empty lines.

If you write proper markdown (use a linter) it's going to work.

**Reply by younes-io** [reply]
@mcanouil Thank you for pointing that out. Could you please recommend any good Markdown linters for VSCode ? (I have already installed an MD linter, but it didn't show any errors or warnings..)

Also, someone suggested that I should be using the CDN Tailwind:

```
format:
  html:
    include-in-header:
      - text: |
          <script src="https://cdn.tailwindcss.com/3.4.1"></script>
```

When I used this, the styles had their effect!

**Reply by mcanouil** [reply]
Well, for the divs, it's not plain markdown, so there is not rules yet in markdown linter.
This being said, the main rule in markdown is that any block elements should be surrounded by empty lines, so just apply the same rule everywhere.

Using the above rule, raw inline, closing divs give me the below result:
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/18f8417f-d870-4f34-8074-03ef9e27c059">

**Reply by younes-io** [reply]
Oh, I see.. you add the `:` for nested divs.. Thank you, I'll try to look into the docs to grasp this very well.

For the styling part, this is the expected render; now it works
![image](https://github.com/quarto-dev/quarto-cli/assets/3153107/b249c975-4e48-4ac8-9fab-f7a2140c38d9)

**Reply by mcanouil** [reply]
The additional `:` is just a trick to see the nested level.
Pandoc does not care much as long as the divs open and close with at least three. Pandoc also does not care if the div open with six but close with three.

**Reply by mcanouil** [reply]
Regarding the styling, I don't know the CDN you are using but following Tailwind CSS official documentation (<https://tailwindcss.com/docs/installation/play-cdn>):

````qmd
---
title: "Fake app"
pagetitle: "Fake page title"
subtitle: "Fake role"
format: html
include-in-header:
  text: |
    <script src="https://cdn.tailwindcss.com"></script>
---

```{=html}
<button class="px-4 py-1 text-sm text-purple-600 font-semibold rounded-full border border-purple-200 hover:text-white hover:bg-purple-600 hover:border-transparent focus:outline-none focus:ring-2 focus:ring-purple-600 focus:ring-offset-2">Message</button>
```
````

<img width="268" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/74dc0a51-cd04-414d-ad30-f89b985729d2">


Note that this has nothing to do with Quarto per se.

**Reply by mcanouil** [reply, marked as answer]
Here when following third party documentation:

<table>
<tr><th>Quarto document</th><th>HTML</th></tr>
<tr>
<td>

````qmd
---
title: "Fake app"
pagetitle: "Fake page title"
subtitle: "Fake role"
format: html
include-in-header:
  text: |
    <script src="https://cdn.tailwindcss.com"></script>
---


::: {.py-8 .px-8 .max-w-sm .mx-auto .bg-white .rounded-xl
  .shadow-lg .space-y-2 .sm:py-4 .sm:flex .sm:items-center
  .sm:space-y-0 .sm:space-x-6}

```{=html}
<img
  class="block mx-auto h-24 rounded-full sm:mx-0 sm:shrink-0"
  src="https://placehold.co/600x400.png"
  alt="Woman's Face"
/>
```

:::: {.text-center .space-y-2 .sm:text-left}

::::: {.space-y-0.5}

Erin Lindford

:::::: {.text-lg .text-black .font-semibold}

Product Engineer

::::::: {.text-slate-500 .font-medium}

:::::::

```{=html}
<button
  class="px-4 py-1 text-sm text-purple-600 font-semibold
  rounded-full border border-purple-200 hover:text-white
  hover:bg-purple-600 hover:border-transparent
  focus:outline-none focus:ring-2 focus:ring-purple-600
  focus:ring-offset-2"
>Message</button>
```

::::::
:::::
::::
:::

````

</td>
<td>
<img width="730" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/13c97b74-c0ab-432d-913b-3bc90fca428f">
</td>
</tr>
</table>

**Comment by younes-io** [comment]
@mcanouil : I really struggle to have Tailwind CSS working properly because it conflicts with Quarto's embedded Bootstrap..

Could you please let me know my options?
I want to have (almost full) control on the layouts.. for example, I created an `esj` template with Tailwind classes, but it's not properly rendering.. I spent many hours on this and it seems that it's a lot of CSS work; so I'm wondering if there is a way to reduce all this CSS "debugging" and have a way to either "configure" a different theme (other than the built-in Bootstrap) or completely build my own theme from scratch.. Because, from what it seems in Quarto's docs, a "theme" means a "look and feel" on top of Bootstrap.
The bootstrap links are still generated on the HTML, so I'm just hopelessly seeking advice

**Reply by mcanouil** [reply]
Quarto website/book structure are based and used Bootstrap.
I am not aware of CMS or tools like that offering the ability to use several frameworks with a satisfying result.

So if you want to completely build a website from scratch using tailwind, I am not sure Quarto is the tool you want at least in its current version.

You could try `minimal: true`, but this will also disable several features that use Bootstrap.

**Reply by younes-io** [reply]
okay, thank you

**Comment by mcanouil** [comment]
FYI:
- #8076

**Reply by younes-io** [reply]
yes, I already checked that one

**Reply by mcanouil** [reply]
It's for future readers.

**Reply by SivuyileNzimeni** [reply]
Just here to say thank you for this Q&A post. It has made a project I am working on much faster.
